File: julia/ch1/Ex123.jl
"""
```jldoctest
julia> include("c:/Users/Yohan/IdeaProjects/sicp/julia/ch1/ExXXX.jl")

julia> ExXXX.func(arg)
```
"""
module Ex123

sq(n) = n * n

divides(a, b) = b % a == 0

function find_divisor(n, test)
    if sq(test) > n
        return n
    elseif divides(test, n)
        return test
    else
        return find_divisor(n, next(test))
    end
end

next(n) = n==2 ? 3 : n + 2

smallest_divisor(n) = find_divisor(n, 2)

end

File: haskell/ch1/Ex123.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load ExXXX
-- Run with func_name arg1...

module Ex123 where

sq n = n * n

divides a b = (mod b a) == 0

next n | n == 2 = 3
next n = n + 2

find_divisor n test | sq test > n = n
find_divisor n test | divides test n = test
find_divisor n test = find_divisor n (next test)

smallest_divisor n = find_divisor n 2
